chain/chain/src/chain.rs

```diff
@@ -325,6 +325,69 @@ pub struct OrphanMissingChunks {
     pub ancestor_hash: CryptoHash,
 }
 
+/// Provides view on the current chain state
+/// Both Chain and ChainUpdate implement this trait,
+/// to avoid duplicate functions
+pub trait ChainAccess {
+    fn orphans(&self) -> &OrphanBlockPool;
+    fn blocks_with_missing_chunks(&self) -> &MissingChunksPool<Orphan>;
+    fn chain_store(&self) -> &dyn ChainStoreAccess;
+
+    /// Check if block header is known
+    /// Returns Err(Error) if any error occurs when checking store
+    ///         Ok(Err(BlockKnownError)) if the block header is known
+    ///         Ok(Ok()) otherwise
+    fn check_header_known(
+        &mut self,
+        header: &BlockHeader,
+    ) -> Result<Result<(), BlockKnownError>, Error> {
+        let header_head = self.chain_store().header_head()?;
+        if header.hash() == &header_head.last_block_hash
+            || header.hash() == &header_head.prev_block_hash
+        {
+            return Ok(Err(BlockKnownError::KnownInHeader));
+        }
+        self.check_known_store(header.hash())
+    }
+
+    /// Check if this block is in the store already.
+    /// Returns Err(Error) if any error occurs when checking store
+    ///         Ok(Err(BlockKnownError)) if the block is in the store
+    ///         Ok(Ok()) otherwise
+    fn check_known_store(
+        &self,
+        block_hash: &CryptoHash,
+    ) -> Result<Result<(), BlockKnownError>, Error> {
+        match self.chain_store().block_exists(block_hash)? {
+            true => Ok(Err(BlockKnownError::KnownInStore)),
+            false => {
+                // Not yet processed this block, we can proceed.
+                Ok(Ok(()))
+            }
+        }
+    }
+
+    /// Check if block is known: head, orphan or in store.
+    /// Returns Err(Error) if any error occurs when checking store
+    ///         Ok(Err(BlockKnownError)) if the block is known
+    ///         Ok(Ok()) otherwise
+    fn check_known(&self, block_hash: &CryptoHash) -> Result<Result<(), BlockKnownError>, Error> {
+        let head = self.chain_store().head()?;
+        // Quick in-memory check for fast-reject any block handled recently.
+        if block_hash == &head.last_block_hash || block_hash == &head.prev_block_hash {
+            return Ok(Err(BlockKnownError::KnownInHead));
+        }
+        // Check if this block is in the set of known orphans.
+        if self.orphans().contains(block_hash) {
+            return Ok(Err(BlockKnownError::KnownInOrphan));
+        }
+        if self.blocks_with_missing_chunks().contains(block_hash) {
+            return Ok(Err(BlockKnownError::KnownInMissingChunks));
+        }
+        self.check_known_store(block_hash)
+    }
```

Comment:
Non-trivial default method impls are a bit of an anti-pattern. As written, an impl of `ChainAccess` would be able to override this logic, although it really shouldn't. Slightly more purist way to write this would be to use a free-standing function: 

```rust
pub fn check_known(chan: &dyn ChainAccess, block_hash: &CryptoHash) -> Result<Result<(), BlockKnownError>, Error> {
   ...
}
```

But this doesn't really matter much.